.circleci/config.yml

```diff
@@ -1,6 +1,48 @@
 version: 2
 
 jobs:
+  compile-upload-tools:
+    parameters:
+      os:
+        type: string
+      tool:
+        type: string
+
+    docker:
+      - image: cimg/python:3.10
+      - environment:
+          OS=<< parameters.os >>
+          # TODO: how do we set this for .org and/or .com?
+          AWS_REGION=
+          AWS_ACCESS_KEY_ID=
+          AWS_SECRET_ACCESS_KEY=
+          AWS_BUILD_TOOLS_BUCKET=
+
+    steps:
+      - checkout
+      - setup_remote_docker:
+          version: 20.10.11
+          docker_layer_caching: true
+      - run:
+          name: Install depedencies
+          command: |
+            pip install awscli==1.20.34
+      - run:
+          name: Download Docker image
+          command: |
+            docker pull readthedocs/build:<< parameters.os >>
+      - run:
+          name: Download compile_version_upload_s3.sh script
+          command: |
+            wget https://raw.githubusercontent.com/readthedocs/readthedocs.org/rel/scripts/compile_version_upload_s3.sh
+      - run:
+          name: Compile tool version
+          command: |
+            # TODO: How do we iterate over `parameterts.tool` properly here?
+            # TODO: How do we run only the command we want, avoiding compiling all the tools/versions each time?
+            ./compile_version_upload_s3.sh << parameters.tool >> << parameters.tool.version >>
```

Comment:
It would probably be easier if the compile.sh script didn't try to do everything at once -- build the image and s3 upload. If it instead just built the tool tarballs, and another script uploaded to s3, the local development and automation here might be easier.

Because the script is trying to do both tasks at once, you're going to need to build all of the tool tarballs twice -- build once for community and upload to community s3, then build the same tarballs for commercial and upload to commercial s3.

Instead, our circle compile job could use one job to compile the tarballs, and then two more jobs strictly to upload the pre-built tarballs -- one to upload to community s3, one to upload to commercial s3.

I did something similar with attached workspaces here: https://github.com/readthedocs/test-sphinx-theme/blob/main/.circleci/config.yml

Reply:
I understand the benefit of splitting build and upload. Basically, we won't build the same tool/version twice (one for commercial and one for community); but that's an optimization. However, I don't understand how that will solve the problem "build and upload _only_ the version(s) that was/were added in this PR"

Reply:
How concerned are you of this scenario? I suppose I'm not too concerned about rebuilding the binary packages. We can waste some resources rebuilding and we have pretty explicit versions defined. It seems we should expect repeatable builds given we're defining:

- Stable releases of system packages by installing LTS release packages
- A specific version of asdf
- Specific versions of binaries to build via asdf

Are there other dependencies you're worried about?

Reply:
I'm sure we are going to hit issues rebuilding each version every time:

- some underlying library required for compiling has changed in the Ubuntu image. Even if they are LTS versions
- conda installer does weird things --like updating some package to latest versions; so they won't be exactly the same on each build
- transitive dependencies for the things _we_ install in the Python versions

We've were here already when re-building old Docker images (e.g. `latest`). At that time, we pinned our own Python dependencies (e.g. `virtualenv`, `setuptools`, etc). Now, it's not the exact same case because we are just re-building tools/versions, but it's a similar situation.

I'm fine testing this if you like the approach. However, we have to know that we could hit issues that won't be _super_ easy to revert.

Reply:
> some underlying library required for compiling has changed in the Ubuntu image. Even if they are LTS versions

The Debian/Ubuntu release process is very deliberate, and LTS packages only see security fixes after LTS release. So, this one doesn't bother me much at all.

> conda installer does weird things

This is a great point, not sure a great way to avoid this besides spacing out our builds of conda and explicitly updating it manually.

I suppose, if we're looking to deduplicate built tarballs, we only really need to query S3 to see if a versioned tarball exists. If the tarball is missing, then we rebuild. That probably solves your concerns, though I'd also be fine calling this extra credit if we need to.

> transitive dependencies for the things we install in the Python versions

I wonder if we should get fancier with our own dependencies, and install from a lock file or pip-tools instead of:

https://github.com/readthedocs/readthedocs.org/blob/main/scripts/compile_version_upload_s3.sh#L115-L130